File: 6_tree/二分搜索树BST/二分搜索树与链表/1586. 二叉搜索树迭代器 II.ts
import { BinaryTree } from '../../力扣加加/Tree'

//
class BSTIterator {
  private stack: BinaryTree[]
  private dummy: BinaryTree
  private cur: BinaryTree | null | undefined

  // 使用树中不存在的最小值来初始化内部指针
  constructor(private root: BinaryTree | null) {
    this.stack = []
    this.dummy = new BinaryTree(Infinity)
    this.cur = this.dummy
    this.pushLeft(root)
  }

  // 游标后面有（中序遍历序列的）节点，或者栈非空（游标节点有父节点）
  hasNext(): boolean {
    return this.cur?.right != null || this.stack.length > 0
  }

  // 从栈中取下一个节点 并对next的right节点执行pushLeft操作 然后删除next的right节点
  next(): number {
    if (this.cur?.right == null) {
      const next = this.stack.pop()!
      this.pushLeft(next.right)
      next.right = null // 这句很关键

      // next和cur串在一起  类似中序遍历转链表常用的pre 和 root
      next.left = this.cur!
      this.cur!.right = next
    }
    this.cur = this.cur?.right
    return this.cur!.val
  }

  hasPrev(): boolean {
    return this.cur !== this.dummy && this.cur?.left !== this.dummy
  }

  prev(): number {
    this.cur = this.cur?.left
    return this.cur!.val
  }

  private pushLeft(root: BinaryTree | null) {
    while (root) {
      this.stack.push(root)
      root = root.left
    }
  }
}

export {}
// 你可以假设 next() 和 prev() 的调用总是有效的

File: 22_专题/迭代器/1586. 二叉搜索树迭代器 II.go
// 1586. 二叉搜索树迭代器 II
// https://leetcode.cn/problems/binary-search-tree-iterator-ii/solutions/2387842/er-cha-sou-suo-shu-die-dai-qi-ii-by-leet-3o49/

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type BSTIterator struct {
	pre     *TreeNode
	stack   []*TreeNode
	arr     []int
	pointer int
}

func Constructor(root *TreeNode) BSTIterator {
	return BSTIterator{
		pre:     root,
		pointer: -1,
	}
}

func (this *BSTIterator) HasNext() bool {
	return len(this.stack) > 0 || this.pre != nil || this.pointer < len(this.arr)-1
}

func (this *BSTIterator) Next() int {
	this.pointer++
	if this.pointer == len(this.arr) {
		for this.pre != nil {
			this.stack = append(this.stack, this.pre)
			this.pre = this.pre.Left
		}
		cur := this.stack[len(this.stack)-1]
		this.stack = this.stack[:len(this.stack)-1]
		this.pre = cur.Right
		this.arr = append(this.arr, cur.Val)
	}
	return this.arr[this.pointer]
}

func (this *BSTIterator) HasPrev() bool {
	return this.pointer > 0
}

func (this *BSTIterator) Prev() int {
	this.pointer--
	return this.arr[max(0, this.pointer)]
}

/**
* Your BSTIterator object will be instantiated and called as such:
* obj := Constructor(root);
* param_1 := obj.HasNext();
* param_2 := obj.Next();
* param_3 := obj.HasPrev();
* param_4 := obj.Prev();
 */
